main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -28,301 +28,346 @@ import ca.uwaterloo.flix.util.Validation._
  */
 object Inliner {
 
+  sealed trait Expression
+
+  object Expression {
```

Comment:
While inlining we meet a problem as sometimes we want to inline a exp that has been visited and sometimes we wanted to visit an exp after it has been inlined. This caused problems with the substituion map as visitexp now returned a different kind of expression 

 It also helped us fix an issue we had previously where would visit a variable that occured Many times on each occurence as we were inlining instead of just once before inlining it. 
 
 There might be a better solution than this

Reply:
OK, I think that makes sense.